onnxruntime/core/graph/contrib_ops/quantization_defs.cc

```diff
@@ -1132,88 +1132,225 @@ If mask is provided, mask index (that is position of first 0 in mask, or number
-                                        AttributeProto::INT, static_cast<int64_t>(1))
-                                  .Input(0, "A", "3-dimensional matrix A", "Q")
-                                  .Input(1, "scale_A", "scale of the input A", "S")
-                                  .Input(2, "B", "2-dimensional matrix B", "Q")
-                                  .Input(3, "scale_B", "scale of the input B", "S")
-                                  .Input(4, "scale_Y", "scale of the output Y", "S")
-                                  .Input(5, "bias", "1d bias", "S", OpSchema::Optional)
-                                  .Input(6, "C",
-                                         "3d or 2d matrix C. if 2d expand to 3d first. "
-                                         "Shape[0] should be 1 or same as A.shape[0] ",
-                                         "Q", OpSchema::Optional)
-                                  .Input(7, "scale_C", "scale of the input A", "S", OpSchema::Optional)
-                                  .Output(0, "Y", "Matrix multiply results from A * B", "Q")
-                                  .TypeConstraint("Q", {"tensor(int8)"},
-                                                  "Constrain input and output types to int8 tensors.")
-                                  .TypeConstraint("S", {"tensor(float)"}, "Constrain bias and scales to float32")
-                                  .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-                                    propagateElemTypeFromInputToOutput(ctx, 0, 0);
-                                    ONNX_NAMESPACE::matmulShapeInference(ctx, 0, 2);
-                                  }));
+  ONNX_MS_OPERATOR_SET_SCHEMA(
+      QuantizeWithOrder,
+      1,
+      OpSchema()
+          .SetDoc(R"DOC(Quantize input matrix to specific layout used in cublaslt.)DOC")
+          .Attr("order_input", "cublasLt order of input matrix", AttributeProto::INT)
+          .Attr("order_output", "cublasLt order of output matrix", AttributeProto::INT)
+          .Input(0, "input", "TODO: input tensor of (ROWS, COLS). if less than 2d, will broadcast to (1, X). If 3d, it is treated as (B, ROWS, COS)", "F")
+          .Input(1, "scale_input", "scale of the input", "S")
+          .Output(0, "output", "output tensor", "Q")
+          .TypeConstraint("Q", {"tensor(int8)"}, "Constrain input and output types to int8 tensors.")
+          .TypeConstraint("F", {"tensor(float16)", "tensor(float)"}, "Constrain to float types")
+          .TypeConstraint("S", {"tensor(float)"}, "Constrain Scale to float32 types")
+          .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+            propagateElemTypeFromDtypeToOutput(ctx, ONNX_NAMESPACE::TensorProto::INT8, 0);
+            if (!hasInputShape(ctx, 0)) return;
+            auto& input_shape = getInputShape(ctx, 0);
+            updateOutputShape(ctx, 0, input_shape);
```

Comment:
I guess output_shape depends on the order_input adn the order_output and not always the same as the input_shape ? 
Btw, what is the "right" output_shape if the order_output is not ROW_MAJOR ? It seems like we could have an ORDER_COL output with ORDER_ROW input's shape and this maybe lead to problems while debugging issues...

Reply:
Literally they have same shape, although different layout/order. Some unsupported case is that maybe we need padding somewhere, or remove padding somewhere. Yet currently those case are not considered.